RachelWilliamsDev/SwiftList
# feat: clerk middleware redirects and organisations

### Description

This PR introduces middleware and UI changes to implement proper redirection for public and protected routes in our application using Clerk. It also includes the ability to create organizations with Clerk and integrates an OrganizationSwitcher into the navigation bar.

### 📸 Screenshots

**Organization Creation**

<img width="1620" alt="Screenshot 2024-12-24 at 10 12 12" src="https://github.com/user-attachments/assets/b7f12eb8-2357-43c8-b840-f9ea7955247c" />

**Organization Switcher in Navbar**

<img width="1624" alt="Screenshot 2024-12-24 at 10 13 05" src="https://github.com/user-attachments/assets/d6df4798-2ea4-4c3b-97fe-c42d4912aed0" />


### ✨ Key Changes

1. Clerk Middleware

- Added logic to redirect users to the sign-in page if they're not authenticated and try to access protected routes.
- Redirects users with no organization selected to /select-org.

2. Routing

- Configured route matchers for public and protected routes.
- Added fallback redirection logic to /select-org or /organization/[orgId] based on orgId.

3. UI Updates

- Integrated OrganizationSwitcher in the navbar for seamless organization management.
- Created select-org page for organization selection.

4. API Integration

- Added API calls and state management for Clerk's organization features.

### ✅ Checklist

- [X] Middleware logic handles all defined scenarios.
- [X] Routes redirect appropriately for unauthenticated users.
- [X] OrganizationSwitcher works seamlessly in the navbar.
- [X] Unit tests for middleware and UI components are added and passing.
- [X] Manual tests for public/protected routes, organization selection, and switching.

### 🚀 How to Test

1. Test Clerk:

- Log in with a user that has and does not have an organization.

2. Test Routes:

- Navigate to public routes and verify access is granted.
- Navigate to protected routes without logging in and confirm redirection to /sign-in.
- After logging in, navigate to /select-org and select an organization.

3. Navbar Organization Switcher:

- Verify the organization switcher displays correctly.
- Switch between organizations and confirm proper state updates.

4. Edge Cases:

- Ensure accessing /dashboard with no organization redirects to /select-org.
- Confirm invalid routes do not break the middleware.

### 🚧 Future Enhancements
